---
title: "Assignment 9"
author: "Gabi Carley"
date: "November 6, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(devtools)
library(rafalib)
library(RColorBrewer)
library(qvalue)
library(genefilter)
library(limma)
library(Biobase) 
library(GSE5859)
```
# Exercise 1 

###Load the admissions data from the dagdata package (which is available from the genomicsclass repository):
```{r}
library(dagdata) 
data(admissions)
```

###Familiarize yourself with this table:
```{r}
admissions
```

###1. Let's compute the proportion of men who were accepted:
```{r}
index = which(admissions$Gender==1) 
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100) 
applied = sum(admissions$Number[index]) 
accepted/applied
```

###What is the proportion of women that were accepted? 
```{r}
index = which(admissions$Gender==0)
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100)
applied = sum(admissions$Number[index])
accepted/applied
```

###2. Now that we have observed different acceptance rates between genders, test for the significance of this result. If you perform an independence test, what is the p-value?
```{r}
atab<-as.data.frame(matrix(0,2,2))
rownames(atab)<-c("Men","Women")
colnames(atab)<-c("Accepted","Rejected")
index = admissions$Gender==1
men=admissions[index,]
women=admissions[!index,]
atab[1,1]= sum(men$Number * men$Percent/100)
atab[1,2]= sum(men$Number*(1-men$Percent/100))
windex = which(admissions$Gender==0)
atab[2,1]= sum(women$Number * women$Percent/100)
atab[2,2]= sum(women$Number * (1-women$Percent/100))
atab
chisq.test(atab)$p.value
```
The p-value is 9.139492e-22.
###This difference actually led to a lawsuit???. Now notice that looking at the data by major, the differences disappear.
```{r}
index = admissions$Gender==1
men = admissions[index,]
women = admissions[!index,]
print( data.frame( major=admissions[1:6,1],men=men[,3], women=women[,3]) )
```

###How can this be? This is referred to as Simpson's Paradox. In the following questions we will try to decipher why this is happening. 

###3. We can quantify how "hard" a major is by using the percent of students that were accepted. Compute the percent that were accepted (regardless of gender) to each major and call this vector H. Which is the hardest major? 
```{r}
major = admissions[1:6,1]
men = admissions[1:6,]
women =admissions[7:12,]
H = (men$Number*men$Percent/100 + women$Number*women$Percent/100) / (men$Number+women$Number)
H
major[which.min(H)]
min(H)
```
Major F is the hardest.
###4. What proportion is accepted for this major? 
The proportion accepted for this major is 0.06477591.

###5. For men, what is the correlation between the number of applications across majors and H? 
```{r}
cor(H,men$Number)
```
The correlation for men between the number of applications across majors and H is 0.7647567.
###6. For women, what is the correlation between the number of applications across majors and H? 
```{r}
cor(H,women$Number)
```
The correlation for women between the number of applications across majors and H is -0.6743393.
###7. Given the answers to the above, which best explains the differences in admission percentages when we combine majors? 
. A) We made a coding mistake when computing the overall admissions percentages. 
. B)There were more total number of women applications which made the denominator much bigger. 
. C) There is confounding between gender and preference for "hard" majors: females are more likely to apply to harder majors. 
. D)The sample size for the individual majors was not large enough to draw the correct conclusion. 

The differences in admission percentages can be explained by C- there is confounding between gender and preference for "hard" majors: females are more likely to apply for hard majors.

#Exercise 2

###We will use the Bioconductor package Biobase which you can install with install_bioc function from rafalib: Load the data for this gene expression dataset:
```{r}
library(Biobase) 
library(GSE5859) 
data(GSE5859)
```

###This is the original dataset from which we selected the subset used in GSE5859Subset. We can extract the gene expression data and sample information table using the Bioconductor functions exprs and pData like this:
```{r}
geneExpression = exprs(e) 
sampleInfo = pData(e)
```

###1. Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses, it does appear to have an effect so we will explore this here. You can extract the year from each date like this:
```{r}
year = format(sampleInfo$date,"%y")
```

###Note that ethnic group and year is almost perfectly confounded:
```{r}
table(year,sampleInfo$ethnicity)
```

###2. For how many of these years do we have more than one ethnicity represented? 
```{r}
tab=table(year,sampleInfo$ethnicity)
print(tab)
x=rowSums(tab!=0)
sum(x>=2)
```
More than one ethnicity is represented for two years.
###3. Repeat the above exercise, but now, instead of year, consider the month as well. Specifically, instead of the year variable defined above use:
```{r}
month.year = format(sampleInfo$date,"%m%y")
###For what proportion of these month.year values do we have more than one ethnicity represented? 

mytab=table(month.year,sampleInfo$ethnicity)
print(mytab)
myx=rowSums(mytab!=0)
mean(myx>=2)
```
More than one ethnicity is represented for 0.04761905 of the mont.year values.
###4. Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values < 0.05 ? 
```{r}
library(qvalue)
library(genefilter)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("02","03") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
4308 genes have q-values less than 0.05.
###5. What is the estimate of pi0 provided by qvalue: 
```{r}
qval$pi0
```
The qvalue provides an estimate of 0.3628642 for th pi0.
###6. Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those processed in 2004. Then use the q value package to obtain q-values for each gene. How many genes have q-values less than 0.05?
```{r}
library(qvalue)
library(genefilter)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("03","04") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
2463 genes have q-values less than 0.05.
###7. Now we are going to compare ethnicities as was done in the original publication in which these data were first presented. Use the qvalue function to compare the ASN population to the CEU population. Once again, use the qvalue function to obtain q-values. How many genes have q-values < 0.05 ? 
```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
index = which(ethnicity%in% c("CEU","ASN"))
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
7217 genes have q-values less than 0.05.
###8. Over 80% of genes are called differentially expressed between ethnic groups. However, due to the confounding with processing date, we need to confirm these differences are actually due to ethnicity. This will not be easy due to the almost perfect confounding. However, above we noted that two groups were represented in 2005. Just like we stratified by majors to remove the "major effect" in our admissions example, here we can stratify by year and perform a t-test comparing ASN and CEU, but only for samples processed in 2005. How many genes have q-values < 0.05 ?
```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
index = which(ethnicity%in% c("CEU","ASN") & year=="05")
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
560 genes have q values less than 0.05.
###Notice the dramatic drop in the number of genes with q-value < 0.05 when we fix the year. However, the sample size is much smaller in this latest analysis which means we have less power:
```{r}
table(sampleInfo$ethnicity[index])
```

###9. To provide a more balanced comparison, we repeat the analysis, but now taking 3 random CEU samples from 2002. Repeat the analysis above, but comparing the ASN from 2005 to three random CEU samples from 2002. Set the seed at 3, set.seed(3) How many genes have q-values < 0.05 ?
```{r}
library(qvalue)
library(genefilter)
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
indexASN = which(ethnicity%in% c("ASN") & year=="05")
indexCEU = which(ethnicity%in% c("CEU") & year=="02")
set.seed(3)
indexCEU<-sample(indexCEU,3)
index<-c(indexASN,indexCEU)
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```
3695 genes have q-values less than 0.05.
#Exercise 3 

###For the dataset we have been working with, models do not help due to the almost perfect confounding. This is one reason we created the subset dataset:
```{r}
library(GSE5859Subset) 
data(GSE5859Subset)
```

###Here we purposely confounded month and group (sex), but not completely:
```{r}
sex = sampleInfo$group 
month = factor( format(sampleInfo$date,"%m")) 
table( sampleInfo$group, month)
```

###1. Using the functions rowttests and qvalue compare the two groups. Because this is a smaller dataset which decreases our power, we will use the more lenient FDR cut-off of 10%. How many gene have q-values less than 0.1? 
```{r}
library(qvalue)
library(genefilter)
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.1)
```
59 genes have q values less than 0.1.
###2. Note that sampleInfo$group here presents males and females. Thus, we expect differences to be in on chrY and, for genes that escape inactivation, chrX. We do not expect many autosomal genes to be different between males and females. This gives us an opportunity to evaluate false and true positives with experimental data. For example, we evaluate results using the proportion genes of the list that are on chrX or chrY. For the list calculated above, what proportion of this list is on chrX or chrY? 
```{r}
library(qvalue)
library(genefilter)
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```
The proportion of genes on the list that are on chrX or chrY is 0.3389831.
###3. We can also check how many of the chromosomes X and Y genes we detected as different. How many are on Y? 
```{r}
library(qvalue)
library(genefilter)
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrY")
sum(index)
```
8 of genes detected are on chromosome Y.
###4. Now for the autosomal genes (not on chrX and chrY) for which q-value < 0.1, perform a t-test comparing samples processed in June to those processed in October. What proportion of these have p-values <0.05 ?
```{r}
library(qvalue)
library(genefilter)
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = which(qvals<0.1 & !geneAnnotation$CHR%in%c("chrX","chrY"))
month = factor( format(sampleInfo$date,"%m"))
pval = rowttests(geneExpression[index,], month)$p.value
mean(pval<0.05)
```
The proportion of autosmal genes with q-values less than .1 and processed in June compared to October with p-values less than 0,05 is 0.8717949.
###5. The above result shows that the great majority of the autosomal genes show differences due to processing data.This provides further evidence that confounding isresulting in false positives. So we are going to try to model the month effect to better estimate the sex effect. We are going to use a linear model: Which of the following creates the appropriate design matrix? 
. A) X = model.matrix(???sex+ethnicity) 
. B) X = cbind(sex,as.numeric(month)) 
. C) It can't be done with one line. 
. D) X = model.matrix(???sex+month) 
```{r}
X= model.matrix(~sex+month)
X
```
Anser D, X=model.matrix(~sex+month) creates the appropriate design matrix.
###6. Now use the X defined above, to fit a regression model using lm for each gene.You can obtain p-values for estimated parameters using summary. Here is an example
```{r}
X = model.matrix(~sex+month)
i = 234
y = geneExpression[i,]
fit = lm(y~X-1)
summary(fit)$coef
```

###How many of the q-values for the group comparison are now <0.1? Note the big drop from what we obtained without the correction. 
```{r}
X = model.matrix(~sex+month)
pvals <- t( sapply(1:nrow(geneExpression),function(j){
  y <- geneExpression[j,]
  fit <- lm(y~X-1)
  summary(fit)$coef[2,4] # the 2 picks out the second (sex) coef, the 4 picks out the p value
} ) )
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```
17 of the q-values are now less than 0.1.
###7. With this new list, what proportion of these are chrX and chrY? Notice the big improvement. 
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```
0.8823529 or 88.24% are chrX and chrY.
###8. How many on Y or X? 
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
sum(index)
```
 15 are on Y or X.
###9. Now from the linear model above,extract the p-values related to the coefficient representing the October versus June differences using the same linear model. How many of the q-values for the month comparison are now <0.1? This approach is basically the approach implemented by Combat.
```{r}

X <- model.matrix(~sex+month)
fit <- lmFit(geneExpression,X)
k <- 3 ##third coef, for month
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```
3170 of the q-values for the month comparison are now less than 0.1.
#Exercises 4
```{r}
library(GSE5859Subset)
data(GSE5859Subset)
```
###1. Suppose you want to make an MA plot of the first two samples y= geneExpression[,1:2]. Which of the following projections gives us the projection of y so that column2 versus column 1 is an MA plot?
Answer choice B gives us the projection of y so that column2 versus column1 is an MA plot.

###2. Say Y is MAxN,in theS VDY = UDV^T which of the following is not correct? 
A) DV^T are the new coordinates for the projection U^TY
B) UD are the new coordinates for the projection YV
C) D are the coordinates of the projection U^TY
D) U^TY is a projection from an N-dimensional to M-dimensional subspace.

Answer choice C is not correct. D are not the coordinates for the projection U^TY.

###3. Define: 
```{r}
y = geneExpression - rowMeans(geneExpression)
```

###Compute and plot an image of the correlation for each sample. Make two image plots of these correlations. In the first one, plot the correlation as image. In the second, order the samples by date and then plot an image of the correlation. The only difference in these plots is the order in which the samples are plotted.
###Based on these plots, which of the following you would say is true?
A) The samples appear to be completely independent of each other.
B) Sex seems to be creating structures as evidenced by the two cluster of highly
correlated samples.
C) There appear to be only two factors completely driven by month.
D) The fact that in the plot ordered by month we see two groups mainly driven by
month, and within these we see subgroups driven by date, seems to suggest date more than month per se are the hidden factors.
```{r}
mypar(1,2)
n=ncol(y)
cors=cor(y)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
o <- order(sampleInfo$date)
yo<-y[,o]
cors=cor(yo)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
```
Based on the plots, I think answer choice D- The fact that in the plot ordered by month we see two groups mainly driven by
month, and within these we see subgroups driven by date, seems to suggest date more than month per se are the hidden factors, is correct.

###4. Based on the correlation plots above, we could argue that there are at least two hidden factors. Using PCA estimate these two factors. Specifically, apply the svd to y and use the first two PCs as estimates.
###Which command gives us these estimates?
```{r}
s <- svd(y)
What <- s$v[,1:2] # these are the first two PCs.
rownames(What)<-colnames(y)
round(What,2)
fit = s$u[,1:2]%*% (s$d[1:2]*t(What))
var(as.vector(fit))/var(as.vector(y))
```

###5. Plot each of th eestimated factors ordered by date.Use color to denote month.The first factor is clearly related to date. Which of the following appear to be most different according to this factor?
```{r}
month=format(sampleInfo$date, "%m")
pcs = svd(y)$v[,1:2]
o = order(sampleInfo$date)
cols = as.numeric(month)[o]
mypar(2,1)
for(i in 1:2){
  plot(pcs[o,i],col=cols,xaxt="n",xlab="")
  label = gsub("2005-","",sampleInfo$date[o])
  axis(1,1:ncol(y),label,las=2)
}
```
June 23 and June 27 appear to be the most different.
###6. Use the svd function to obtain the principal components (PCs) for our detrended gene expression data y. How many PCs explain more than 10% of the variability?
```{r}
s = svd(y)
varexplained = s$d^2/ sum(s$d^2)
plot(varexplained)
sum(varexplained>0.10)
```
2 PCs account for more than 10% of the variability.
###7. Which PC most correlates (negative or positive correlation) with month?
```{r}
s = svd(y)
month = factor( format(sampleInfo$date,"%m"))
cors = cor( as.numeric(month),s$v)
plot(t(cors))
which.max(abs(cors))
max(abs(cors))
```
PC 1 most correlates with month.
###8. What is this correlation (in absolute value)?
The correlation is 0.8297915.

###9. Which PC most correlates (negative or positive correlation) with sex?
```{r}
s = svd(y)
sex = sampleInfo$group
cors = cor( as.numeric(sex),s$v)
plot(t(cors))
which.max(abs(cors))
max(abs(cors))
```
PC 1 also correlates most with 6.
###10. What is this correlation (in absolute value)?
The correlation is 0.6236858.

###11. Now instead of using month, which we have shown does not quite describe the batch, add the two estimated factors s$v[,1:2] to the linear model we used above. Apply this model to each gene and compute q-values for the sex difference. How many q-values < 0.1 for the sex comparison?
```{r}
library(limma)
X <- model.matrix(~sex+s$v[,1:2])
fit <- lmFit(y,X)
k <- 2 ##second coef, for month
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```
There are 14 q-values less than 0.1 for the sex  comparsion.
###12. What proportion of the genes are on chromosomes X and Y?
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```
All of the genes are on chromosomes x and y.
#Exercises 5
```{r}
library(GSE5859Subset)
data(GSE5859Subset)
library(sva)
```
###1. In a previous section we estimated factors using PCA, but we noted that the first factor was correlated with our outcome of interest:
```{r}
    s <- svd(geneExpression-rowMeans(geneExpression))
    cor(sampleInfo$group,s$v[,1])
```
###The svafit function estimates factors, but downweighs the genes that appear to correlate with the outcome of interest. It also tries to estimate the number of factors and returns the estimated factors like this:
```{r}
    sex = sampleInfo$group
    mod = model.matrix(~sex)
    svafit = sva(geneExpression,mod)
    head(svafit$sv)
```
###The resulting estimated factors are not that different from the PCs.
```{r}
for(i in 1:ncol(svafit$sv)){
print( cor(s$v[,i],svafit$sv[,i]) ) }
```
###Now fit a linear model to each gene that instead of month includes these factors in the model. Use the qvalue function. How many genes have q-value < 0.1?
```{r}
svaX<-model.matrix(~sex+svafit$sv)
library(limma)
fit <- lmFit(geneExpression,svaX)
k <- 2 #second coef, for sex
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
library(qvalue)
library(genefilter)
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```
13 genes have a q-value of less than 0.1.
###2. How many of these genes are from chrY or chrX?
```{r}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
sum(index)
```
The proportion of genes from chrY or chrX is 0.9230769, or 12 genes.